### PR TITLE
Accept a Redact function

### LINKED ISSUE #5: Possible to redact partially?
Is it possible to redact partially, like the way credit card numbers or SSN number are done?

E.g.
```js
{ secret: "1234567890" } 

// outputs

{ secret: "xxxxxx7890" }
```

This will be very helpful to identify potential issues without delving the entire secret. 

I think the easiest and most flexible way would be to accept a function? Something like this:

```js
censor = (value) => {
// apply mask
return maskedValue
}
```

I understand this will cause a dip in performance which should be acceptable since this will be opt-in. And if there's a way to do it without sacrificing performance, even better!

Thoughts?

### PR DESCRIPTION
**Pull request by mrchief**
Fixes #5

### CODE REVIEW

**Comment by mrchief** [comment]
@davidmarkclements Here's a simple attempt. I feel like we can do away with the function check everytime and replace it with a simple statement but can't put my finger on how. Also, feel free to add/modify test names, readme changes etc. Hope this helps.

File: readme.md [hunk 1]
```diff
@@ -27,6 +27,8 @@ The `censor` can be of any type, for instance an object like `{redacted: true}`
 is allowed, as is `null`. Explicitly passing `undefined` as the `censor` will
 in most cases cause the property to be stripped from the object. Edge cases occur when an array key is redacted, in which case `null` will appear in the array (this is ultimately a nuance of `JSON.stringify`, try `JSON.stringify(['a', undefined, 'c'])`)
 
+`censor` can also accept a function. This is helpful in cases where you want to redact dynamically instead of a fixed value. A common use case could be to mask partially (e.g. `{ test: 1234567890 }` => `{ "test": "xxxxxx7890" }`). Check the tests to see this in action.
```
Comment (davidmarkclements) [inline readme.md:30]: let's not tell people to check the tests - feel free to include an example in readme

File: index.js [hunk 2]
```diff
@@ -46,7 +46,7 @@ function noir (serializers, keys, censor) {
   return shape
 
   function redact () { return {toJSON: mask} }
-  function mask () { return censor }
+  function mask () { return typeof censor === 'function' ? censor() : censor }
```
Comment (davidmarkclements) [inline index.js]: the point of `censor` as a function is to take a value and then return a mask based on that value

no value is being passed to `censor` here.. (which could cause user issues as well, if they're expecting a value)

**Review by davidmarkclements — CHANGES_REQUESTED** [review]
couple of things

**Comment by davidmarkclements** [comment]
on top of the changes we also need a comparison benchmark 

easiest way to do this would be to copy bechmarks/index.js - perhaps into 
`benchmarks/static-censor.js` and `benchmarks/functional-censor.js` and modify the files to use static and functional benchmarks

**Comment by davidmarkclements** [comment]
oh could you also please run a comparison benchmark against master and post results? to see the costs of the checking?

File: index.js (hunk 2 above)
Comment (mrchief) [reply index.js]: This is one of the parts I don't have a good understanding of. There is no argument here that I can use. Not sure how to get a value... Can you think of a use case that'll fail this implementation? If you look at my current tests, this only gets called when an empty object is passed. If I can get an understanding, I can create a failing test.

File: readme.md (hunk 1 above)
Comment (mrchief) [reply readme.md:30]: Sure.

File: index.js (hunk 2 above)
Comment (davidmarkclements) [reply index.js]: `mask` becomes the value of `toJSON` for top level objects 

`toJSON` is passed the key, while `this` is the object itself, so you can do `censor(this[key])`

see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify

File: index.js [hunk 3]
```diff
@@ -46,7 +46,7 @@ function noir (serializers, keys, censor) {
   return shape
 
   function redact () { return {toJSON: mask} }
-  function mask () { return censor }
+  function mask (key) { return typeof censor === 'function' ? censor(this[key]) : censor }
```
Comment (davidmarkclements) [inline index.js:49]: can we make sure this logic path had a unit test

Comment (mrchief) [reply index.js:49]: Oh sure. I wasn't done with it. :)

**Comment by yellowbrickc** [comment]
Hi,
can I help somehow to get this pull request merged? It is just was I was looking for!

**Comment by mrchief** [comment]
@yellowbrickc If you can help fix the review comments (adding a readme example, adding a test and benchmarks), it'd be great. I got pulled into other projects and haven't had the chance to revisit this.

**Comment by yellowbrickc** [comment]
I am pretty sure that I cannot fork your changes. The only thing what I can do is to fork the repo, reapply your changes and fill the gaps. Or is there some other way I don't know about?

**Comment by yellowbrickc** [comment]
@mrchief everything is cool, I forked your fork ;)

**Comment by mrchief** [comment]
Oh Ok. I added you as collaborator to my fork also. If you push to that, it'll update this PR. If not, you'll need to create a new PR from your fork.

File: readme.md (hunk 1 above)
Comment (yellowbrickc) [reply readme.md:30]: @davidmarkclements can you please review the pull request? I added a benchmark run and an example to the readme.

Comment (yellowbrickc) [reply readme.md:30]: sorry, I just saw you would like a special benchmark. I didn't saw that, sorry. Maybe we can live with the simple comparison I did?

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "pino-noir",
-  "version": "2.0.0",
+  "version": "2.0.1",
```
Comment (davidmarkclements) [inline package.json]: this should be a minor increase not a patch, since it's an API extension

**Comment by davidmarkclements** [comment]
hey @yellowbrickc thanks for picking this up!

benchmark looks good, could you please run the benchmark and post the results here? 

we also need to add a test for line 49 in index.js

**Review by davidmarkclements — CHANGES_REQUESTED** [review]

File: package.json (hunk 4 above)
Comment (yellowbrickc) [reply package.json]: can you give me a hint here: how do I run the benchmark? I looked for a command but didn't found it anywhere :(
I would include this as a script tag afterwards.

Comment (yellowbrickc) [reply package.json]: Regarding the tests: there is one called "masks according to supplied censor function" which covers the new functionality and and another one "masks according to supplied censor" covering the old functionality.
The coverage report displays only line 11 as uncovered:
index.js |      100 |    98.04 |      100 |      100 |             11 |

Comment (mcollina) [reply package.json]: @yellowbrickc you can run the benchmarks with `node benchmarks`. You can also add that to the package.json if you like.

File: package.json [hunk 5]
```diff
@@ -9,7 +9,8 @@
   "scripts": {
     "test": "standard && tap test",
     "cov-badge": "tap --coverage-report=lcov && codecov -t 346eeb5b-38b7-4ff5-b0ec-eb2da6a1254f",
-    "cov": "tap test --cov"
+    "cov": "tap test --cov",
+    "benchmark": "node benchmarks ."
```
Comment (davidmarkclements) [inline package.json]: there's no `.` required - `benchmarks` is the path

**Comment by davidmarkclements** [comment]
great work @yellowbrickc  - one nit pick in `package.json`

thanks for confirming on the test coverage, 

all we need now is to see any performance impact, could you please run the benchmarks and then post the output in this PR thread?

**Comment by yellowbrickc** [comment]
Sure, I can do this tomorrow morning. 
I was wondering about 3 things, maybe you could give me some insights: 
- you are using exclusively vars. As i understand how the linking is happening don't you think that const and let would be better?
- Second question: why do you never finish a function or expression with a semi-colon ?
- Third question: on which node-version should the benchmark run?

**Comment by davidmarkclements** [comment]
hey @yellowbrickc 

* this was first written when const wasn't supported extensively, I'd like to keep with var for certain libraries – pino-noir is for use with pino, and pino still has a legacy branch for Node 0.10.  As for `let` I tend to ignore it, from practice it seems that most times when block scoping is useful, so is immutable binding (`const`) (also out of habit, `var` used to be (may still be) faster than `let`.
* all of the official pino ecosystem modules use [Standard](http://npm.im/standard) as the linter – no semi-colons required
* latest is fine :)  (8.9.4)

**Comment by mcollina** [comment]
Using let generate slower code than var. This is also true for const. Future V8 versions will fix this.

**Comment by yellowbrickc** [comment]
Thanks for the insights :)

Latest benchmark:

benchPinoTopLevel*10000: 333.779ms
benchNoirTopLevel*10000: 382.708ms
benchPinoNested*10000: 491.312ms
benchNoirNested*10000: 481.082ms
benchPinoDeepNested*10000: 467.695ms
benchNoirDeepNested*10000: 500.071ms
benchPinoVeryDeepNested*10000: 518.262ms
benchNoirVeryDeepNested*10000: 634.933ms
benchPinoWildcardStructure*10000: 744.625ms
benchNoirWildcards*10000: 610.159ms
benchPinoFunctionCensor*10000: 334.572ms
benchNoirFunctionCensor*10000: 459.203ms
benchPinoTopLevel*10000: 265.533ms
benchNoirTopLevel*10000: 329.132ms
benchPinoNested*10000: 397.983ms
benchNoirNested*10000: 401.191ms
benchPinoDeepNested*10000: 397.944ms
benchNoirDeepNested*10000: 489.064ms
benchPinoVeryDeepNested*10000: 509.982ms
benchNoirVeryDeepNested*10000: 643.705ms
benchPinoWildcardStructure*10000: 582.878ms
benchNoirWildcards*10000: 564.312ms
benchPinoFunctionCensor*10000: 328.049ms
benchNoirFunctionCensor*10000: 424.109ms

**Review by mcollina — APPROVED** [review]
LGTM

**Comment by yellowbrickc** [comment]
Hi @davidmarkclements 
do you have an idea when you will merge and release this pull request?

**Comment by davidmarkclements** [comment]
Sorry! This fell of my radar - hopefully tomorrow or thu at latest

**Review by davidmarkclements — APPROVED** [review]

**Comment by davidmarkclements** [comment]
released v2.2.0 

really great collab. @yellowbrickc and @mrchief

**Pull request merged**